Sources/AsyncHTTPClient/Utils.swift

```diff
@@ -47,21 +53,85 @@ public final class HTTPClientCopyingDelegate: HTTPClientResponseDelegate {
 }
 
 extension ClientBootstrap {
-    static func makeHTTPClientBootstrapBase(group: EventLoopGroup, host: String, port: Int, configuration: HTTPClient.Configuration, channelInitializer: ((Channel) -> EventLoopFuture<Void>)? = nil) -> ClientBootstrap {
-        return ClientBootstrap(group: group)
-            .channelOption(ChannelOptions.socket(SocketOptionLevel(IPPROTO_TCP), TCP_NODELAY), value: 1)
+    fileprivate func makeClientTCPBootstrap(
+        host: String,
+        requiresTLS: Bool,
+        configuration: HTTPClient.Configuration
+    ) throws -> NIOClientTCPBootstrap {
+        let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
+        let sslContext = try NIOSSLContext(configuration: tlsConfiguration)
+        let hostname = (!requiresTLS || host.isIPAddress) ? nil : host
+        let tlsProvider = try NIOSSLClientTLSProvider<ClientBootstrap>(context: sslContext, serverHostname: hostname)
+        return NIOClientTCPBootstrap(self, tls: tlsProvider)
+    }
+}
 
+extension NIOClientTCPBootstrap {
+    
+    /// create a TCP Bootstrap based off what type of `EventLoop` has been passed to the function.
+    fileprivate static func makeBootstrap(
+        on eventLoop: EventLoop,
+        host: String,
+        requiresTLS: Bool,
+        configuration: HTTPClient.Configuration
+    ) throws -> NIOClientTCPBootstrap {
+        let bootstrap: NIOClientTCPBootstrap
+        #if canImport(Network)
+        // if eventLoop is compatible with NIOTransportServices create a NIOTSConnectionBootstrap
+        if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *),
+            var tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
+            
+            tsBootstrap = tsBootstrap.channelOption(NIOTSChannelOptions.waitForActivity, value: false)
```

Comment:
Why are we setting this to `false` unconditionally?

Reply:
Thinking about this, I'm wondering if there is a case for a change in NIOTS. The reason I set it to `false` was DNS errors were not being returned, I would get a timeout instead. But if you are getting a DNS error I assume you havent reached the server and NIOTS should close the connection at that point instead of waiting.

Reply:
We can consider trying to provide a more expressive timeout error in NIOTS, but in general I think waiting for connectivity is the right choice. As I noted in our discussion over on the NIOTS repository, DNS errors are not final: they may change as network state changes.

Reply:
That isn't what NIOSSL does though, connection errors due to DNS return almost immediately

Reply:
Regular NIO doesn't do it because it can't: it doesn't have sufficient insight into the network to know. In principle it will start to do things like this, though, because if we add more support for streaming service discovery mechanisms those mechanisms may not necessarily have a definitive final "this service does not exist" error state, much as DNS does not.

Reply:
Ok so timeout is considered the correct error for a DNS fail?

Reply:
If I revert this back to `true`. I don't get the DNS error as expected, but I also don't get ECONNREFUSED when I attempt to connect to closed port.

Reply:
> Ok so timeout is considered the correct error for a DNS fail?

It's substantially more complex than that. I think providing more-detailed timeouts would be useful: if we're in `state` `.waiting(error)` then we should provide that error as part of the information when we fire that timeout.

> If I revert this back to true. I don't get the DNS error as expected, but I also don't get ECONNREFUSED when I attempt to connect to a closed port.

Correct, NIOTS hides all the lower-level information. In some ways, though, that information is not super useful. The result is just "I couldn't connect, sorry". Again, though, I'm happy for us to add more detail to the error NIOTS returns.

Reply:
I think that would be useful. It could inform whether a user tries to connect again.

Reply:
So the way this is _intended_ to be used is that the user shouldn't try to connect again: they should set a timeout as long as they are willing to wait for a connection. The analogy here is that trying to connect, looking at the error, and maybe trying again is essentially analogous to polling the network. A better model is to try to be reactive: let the OS make decisions about when it is potentially worth retrying a connection attempt, and let it do so.

Reply:
ok, now you need to make NIOSSL do the same :-)
I'll change NIOTS code to reflect this

Reply:
Note that this is not NIOSSL's job: it's _NIO_'s job. And we cannot currently make NIO do this very thoroughly because the OS doesn't tell us these things.

Reply:
removed setting `waitForActivity` to `false`